File: Kafka-Utils-Java/src/main/java/com/yhyr/comsumer/ListenerRebalance.java
package com.yhyr.comsumer;

import org.apache.kafka.clients.consumer.*;
import org.apache.kafka.common.TopicPartition;

import java.util.*;

/**
 * Class Subject: 在客户端监听Group的Rebalance行为;
 *
 * * 在触发Rebalance前 记录当前正在处理各Partition的offset信息; 在Rebalance后记录被重新分配的Partition信息
 *
 * @author yhyr
 * @since 2018/12/30 18:25
 */
public class ListenerRebalance {
    private static Map<TopicPartition, OffsetAndMetadata> currentOffsets = new HashMap<>();

    private static String brokers = "localhost:9092";
    private static String group = "group_test";
    private static String topic = "topic_demo";

    private static KafkaConsumer<String, String> consumer;

    static {
        Properties props = new Properties();
        props.put("bootstrap.servers", brokers);
        props.put("group.id", group);
        props.put("enable.auto.commit", "true");
        props.put("auto.commit.interval.ms", "1000");
        props.put("key.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
        props.put("value.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
        consumer = new KafkaConsumer<>(props);
    }

    private class CustomHandleRebalance implements ConsumerRebalanceListener {
        @Override
        public void onPartitionsRevoked(Collection<TopicPartition> collection) {
            System.out.println(String.format(
                "Before Rebalance, Assignment partitions is: %s; Current each partition's latest offset is: %s",
                collection.toString(), currentOffsets.toString()));
        }

        @Override
        public void onPartitionsAssigned(Collection<TopicPartition> collection) {
            System.out.println(String.format(
                "After Rebalance, Assignment partitions is: %s; Current each partition's latest offset is: %s",
                collection.toString(), currentOffsets.toString()));
        }
    }

    private void consumer() {
        try {
            consumer.subscribe(Collections.singletonList(topic), new CustomHandleRebalance());
            while (true) {
                ConsumerRecords<String, String> records = consumer.poll(100);
                for (ConsumerRecord<String, String> record : records) {
                    // record current msg's offset
                    currentOffsets.put(new TopicPartition(record.topic(), record.partition()), new OffsetAndMetadata(
                        record.offset()));
                    // processing msg
                    System.out.println("Processing msg : " + record.toString());
                }
            }
        } catch (Exception e) {
            System.out.println("Unexpected error: " + e.getMessage());
        } finally {
            consumer.commitSync(currentOffsets);
        }
    }

    public static void main(String[] args) {
        ListenerRebalance action = new ListenerRebalance();
        action.consumer();
    }
}

File: Kafka-Utils-Python/consumer/listener_rebalance.py
# -*- coding: utf-8 -*-
"""
Subject: 在客户端监听Group的Rebalance行为
    Tips:
        可以通过自定义Consumer Client_id 来实现Rebalance后Partition的分配结果
        Python Consumer 默认的 Client_id 命名规则为: kafka-python-{version}


@Author YH YR
@Time 2018/12/31 14:13
"""
import time
from kafka import KafkaConsumer, ConsumerRebalanceListener


class CustomHandleRebalance(ConsumerRebalanceListener):
    def on_partitions_revoked(self, revoked):
        print('Before Rebalance, Assignment partitions is: {0}'.format(revoked))

    def on_partitions_assigned(self, assigned):
        print('After Rebalance, Assignment partition is: {0}'.format(assigned))


class ListenerRebalance:
    def __init__(self, broker_list, group_name, topic, enable_auto_commit=True, auto_offset_reset='latest',
                 client_id=None):
        self.broker_list = broker_list
        self.topic = topic
        self.group_name = group_name
        self.enable_auto_commit = enable_auto_commit
        self.auto_offset_reset = auto_offset_reset
        self.client_id = client_id

    def consumer(self, process_msg):
        consumer = KafkaConsumer(group_id=self.group_name, bootstrap_servers=self.broker_list, client_id=self.client_id,
                                 enable_auto_commit=self.enable_auto_commit, auto_offset_reset=self.auto_offset_reset)

        consumer.subscribe(self.topic, listener=CustomHandleRebalance())

        while True:
            consumer_records = consumer.poll(100, max_records=1)
            for partition_info, records in consumer_records.items():
                for record in records:
                    process_msg(record)
                    time.sleep(2)


def print_msg(msg_dic):
    print(msg_dic)


if __name__ == '__main__':
    broker_list = 'localhost:9092'
    group_name = 'group_test'
    topic = ['topic_demo']

    action = ListenerRebalance(broker_list, group_name, topic)
    action.consumer(print_msg)
